src/Morphir/Scala/Feature/Codec.elm

```diff
@@ -209,194 +302,91 @@ composeEncoder ctorName ctorArgs =
-                        Scala.Extract (Scala.NamedMatch (Tuple.first arg |> Name.toCamelCase)) generatorRHS
-                    )
-
-        yeildArgValues =
-            ctorArgs
-                |> List.map
-                    (\( argName, argType ) ->
-                        Scala.ArgValue Nothing (Scala.Variable (argName |> Name.toCamelCase))
-                    )
-
-        yeildExpression =
-            Scala.Apply (Scala.Variable (ctorName |> Name.toTitleCase)) yeildArgValues
-    in
-    if List.isEmpty ctorArgs then
-        ( Scala.NamedMatch (ctorName |> Name.toTitleCase), Scala.Apply (Scala.Variable (ctorName |> Name.toTitleCase)) [] )
-
-    else
-        ( Scala.NamedMatch (ctorName |> Name.toTitleCase)
-        , Scala.ForComp generators yeildExpression
-        )
-
-
 {-|
 
     Get an Encoder reference for a Type
 
 -}
-genEncodeReference : Type ta -> Result Error Scala.Value
-genEncodeReference tpe =
+mapTypeToEncoderReference : Name -> Scala.Path -> List Name -> Type ta -> Result Error Scala.Value
+mapTypeToEncoderReference tpeName tpePath typeParams tpe =
     case tpe of
         Type.Variable _ varName ->
-            Ok (Scala.Variable ("encode" :: varName |> Name.toCamelCase))
+            Scala.Variable ("encode" :: varName |> Name.toCamelCase)
+                |> Ok
 
+        -- Assuming that the encoders for a reference have already been handled. We just have to return the encoder reference
         Type.Reference _ ( packageName, moduleName, typeName ) typeArgs ->
             let
                 scalaPackageName : List String
                 scalaPackageName =
                     packageName ++ moduleName |> List.map (Name.toCamelCase >> String.toLower)
 
+                codecPath : List String
+                codecPath =
+                    List.concat [ scalaPackageName, [ "Codec" ] ]
+
+                encoderName : String
+                encoderName =
+                    "encode" :: typeName |> Name.toCamelCase
+
                 scalaReference : Scala.Value
                 scalaReference =
-                    Scala.Ref
-                        scalaPackageName
-                        ("encode" :: typeName |> Name.toCamelCase)
+                    Scala.Ref codecPath encoderName
             in
-            Ok scalaReference
+            case typeArgs of
+                [] ->
+                    Ok <| scalaReference
+
+                _ ->
+                    typeArgs
+                        |> List.map (\typeArg -> mapTypeToEncoderReference tpeName tpePath typeParams typeArg)
+                        |> ResultList.keepFirstError
+                        |> Result.map (List.map (Scala.ArgValue Nothing))
+                        |> Result.map (Scala.Apply scalaReference)
 
         Type.Tuple a types ->
             let
+                tupleEncoderRef : Scala.Value
+                tupleEncoderRef =
+                    Scala.Ref [ "morphir", "sdk", "tuple", "Codec" ] "encodeTuple"
```

Comment:
I believe this will only work if the tuple has exactly 2 elements. In the Morphir IR tuples can have any number of elements so mapping to an SDK function won't work. We need to generate the encoding logic for these the same way we do for record types (only simpler).

Reply:
And we should probably remove the tuple codecs from `morphir-jvm` since they are misleading.